editor/vscode/src/browser.ts

```diff
@@ -57,12 +61,112 @@ export function activate(context: vscode.ExtensionContext) {
+
+        if (previewPanel) {
+            previewPanel.reveal(vscode.ViewColumn.Beside);
+        } else {
+            // Create and show a new webview
+            previewPanel = vscode.window.createWebviewPanel(
+                'slint-preview',
+                'Slint Preview',
+                vscode.ViewColumn.Beside,
+                { enableScripts: true }
+            );
+            previewPanel.webview.onDidReceiveMessage(
+                async message => {
+                    console.log("MESSAGE ", message);
+                    switch (message.command) {
+                        case 'load_file':
+                            const actual_url = Uri.parse(message.url);
+                            console.log("actual = ", actual_url);
+                            const content = await vscode.workspace.fs.readFile(actual_url);
+                            console.log("content = ", content);
+                            return content;
+                    }
+                },
+                undefined,
+                context.subscriptions
+            );
+            previewPanel.webview.html = getPreviewHtml();
+            previewPanel.onDidDispose(
+                () => { previewPanel = undefined; },
+                undefined,
+                context.subscriptions);
+        }
+
+        previewUrl = ae.document.uri.toString();
+        previewPanel.webview.postMessage({ command: "preview", base_url: ae.document.uri.toString(), content: ae.document.getText() });
     }));
 
     context.subscriptions.push(vscode.commands.registerCommand('slint.reload', async function () {
         statusBar.hide();
         await client.stop();
         startClient(context);
     }));
+
+    vscode.workspace.onDidChangeTextDocument(event => {
+        if (previewPanel && event.document.uri.toString() === previewUrl) {
+            previewPanel.webview.postMessage({
+                command: "preview",
+                base_url: event.document.uri.toString(),
+                content: event.document.getText()
+
+            });
+        }
+    });
+}
+
+function getPreviewHtml(): string {
+    // FIXME this should be bundled in the extension, or we need to change this before the release to the release variant
+    let slint_wasm_interpreter_url = "https://slint-ui.com/snapshots/master/wasm-interpreter/slint_wasm_interpreter.js";
+    return `<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Slint Preview</title>
+    <script type="module">
+    "use strict";
+    import * as slint from '${slint_wasm_interpreter_url}';
+    await slint.default();
+
+    const vscode = acquireVsCodeApi();
+
+    async function load_file(url) {
+        console.log("LOAD ", url);
+        let xxx = await vscode.postMessage({
+                        command: 'load_file',
+                        url: url,
+        });
```

Comment:
Perhaps you can build this using promises.

Each time the compiler requests a file, you allocate a new promise, increase a global request counter and take the value as request id. Then send off the file load request via postMessage where you pass the url and the request is. The promise is stored in a global requests js object as property (request if is key). Then you return from the compiler callback (which is not declared as async function) by returning the promise. The compiler will then await on the promise until it is resolved.

When the response comes back, you can find the promise via the request of and call resolve() on the promise. Then in the next tick the compiler should resume because you resolved the promise.